InputStream, OutputStream

ml-rdv opened the pull request.

ml-rdv committed:
Реализовать 2 метода:
Первый сохраняет объект класса в файл в виде байтов
Второй из файла загружает байты и десериализует их в объект

ml-rdv committed:
Corrections

ml-rdv committed:
Corrections

ml-rdv committed:
Реализовать консольный мини-редактор текстовых файлов. Редактор должен:
При запуске отображать списком файлы и папки текущей папки из которой запускается программа (желательно визуально разделить файлы и папки)
При вводе в консоль команды open c названием папки (пример “open work”)
Должен произойти переход в данную папку. Визуально должен выводиться список её содержимого
Если папки нет, то вывести соответствующее сообщение
При вводе в консоль команды create c названием папки (пример “create documents”)
Должна создаться папка и должен произойти переход в неё
Если папка существует, то вывести соответствующее сообщение
При вводе в консоль команды back происходит возврат на родительскую папку
При вводе в консоль команды open c названием файла (пример “open notes”)
Если файл содержит расширение txt, то содержимое файла выводится в консоль
Если иное расширение, то выводится текст по типу “неверный формат файла”
При нажатие enter содержимое файла убирается и отображается содержимое текущей папки
При вводе в консоль команды create c названием файла (пример “create notes”) создается пустой файл
Если к команде добавить текст, то файл создается с текстом (Пример “create notes hello world”)
При вводе в консоль команды edit с названием файла (Пример “edit notes”)
Можно ввести любой текст и после нажатия enter он добавится в конец файла
При вводе в консоль команды delete и названия файла/папки
Происходит удаление выбранного
Если папка не пустая, то программа должна запросить разрешение на удаление содержимого
Если файла/папки нет, то вывести соответствующее сообщение
При вводе в консоль команды rename, старого и нового названия  файла/папки (пример rename oldName newName) должно произойти переименование
При вводе в консоль команды info и названия файла/папки должна выводится мета информация
Полный(абсолютный) путь
Размер
Дата создания
Если папка, то выводить кол-во вложенных элементов
Желательно перед выполнением команд делать очистку консоли

ml-rdv committed:
Methods are private, except main method
currentPath is deleted
updating directory happens only if user change directory

ml-rdv committed:
The principle of sole responsibility is implemented
Classes ConsoleTextFileEditor and Main are separated
Methods in ConsoleTextFileEditor are public

ml-rdv committed:
Tests for ConsoleTextFileEditor

ml-rdv committed:
Separation of file system manager and work in console

Adgam commented on .idea/artifacts/hw1_2_jar.xml:
папку .idea хорошо добавлять в .gitignore, за редким исключением некоторые настройки в папке .idea распространяются на команду, например конфигурации запуска приложений, но в основном папка игнорируется

ml-rdv replied:
> @Adgam fixed

Adgam commented on SerializedObject:
Подобные файлы не нужно пушить. Бинарный файлы, скомпилированные файлы, результаты выполнения программ и тяжёлые медиафайлы желательно не пушить, т.к. засоряют репозиторий, особенно, если они много весят.  Проблема ещё в том, что даже после удаления они остаются в истории гита и всё также занимают место, удалить их окончательно можно, но проблематично. 

Если случайно запушила, то можно сделать в репозитории папку по типу `work` или `tmp` и добавить её в .gitignore, а в программах соответственно работать с этой папкой.

ml-rdv replied:
> @Adgam fixed

Adgam commented on src/META-INF/MANIFEST.MF:
Можно было бы сделать jar через IDEA тогда не нужно добавлять манифест руками. 
В данном случае, я бы тоже его добавил в .gitignore, но допускаю, что в случаях более сложных может быть необходимость хранить манифест в репе и поддерживать его версионность.

ml-rdv replied:
> @Adgam fixed

Adgam commented on src/InputOutputStream/ObjectStream/SerializedObject:
https://github.com/ml-rdv/hw1_2/pull/9/files#r1486152305

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+import java.util.List;
+import java.util.Map;
+
+public class ConsoleTextFileEditorTest {
```
Раздели тесты на 2 класса, 1 для ConsoleTextFileEditor и 1 для FileSystemManagement. Данный тест перегруженным получается и вводит в заблуждение.

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+
+public class ConsoleTextFileEditorTest {
+
+    ConsoleTextFileEditor consoleTextFileEditor;
```
добавить `private final`, даже в тест есть вероятность, что состояние объекта будет изменено, что приведёт к влиянию на тест. Тесты должны быть изолированны как только это возможно.

ml-rdv replied:
@Adgam Зачем делать consoleTextFileEditor final? Тогда ведь нельзя будет в каждом тесте задавать новое значение

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+
+    @BeforeEach
+    public void setEditorAndManager() {
+        consoleTextFileEditor = new ConsoleTextFileEditor();
```
Я бы в каждом тесте вынес создание consoleTextFileEditor и manager явно, т.к. у тебя итак это происходит по факту из-за @BeforeEach, но если включить параллельное выполнение тестов, то из-за того, что у тебя идёт обращение через поля, то тесты будут друг на друга влиять и будут выдавать неправильный результат. Не забывай про изолированность, когда у классов есть состояние.

ml-rdv replied:
@Adgam fixed
Вопрос. Для чего в каждом тесте создавать consoleTextFileEditor? Если в блоке @beforeeach на каждый тест создаётся отдельный объект с собственной ссылкой, как они могут пересекаться при параллельном выполнении тестов?

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    @Test
+    public void directoryMustBeOpened() {
```
Тест небольшой, а проблем много, по порядку:

1. После выделения отдельного класса FileSystemManagement, у ConsoleTextFileEditor по сути должен остаться публичным только метод start. Сейчас класс ConsoleTextFileEditor должен принимать команды из консоли и выводить соответствующий результат в консоль, вот это и нужно нам тестировать. 

2. Метод getCurrentDirectory в ConsoleTextFileEditor нелегальный и сделанный только ради теста, по сути костыль, который никак не связан с обязанностью класса.

3. Лазить в src плохо, т.к. после компиляции jarника этой папки не будет и тесты будут падать с ошибкой. Нужно создать и очищать перед каждым тестов временную рабочую директорию и в ней проводить тесты. Связи с этим в редакторе не хватает гибкой настройки по установки директории по умолчанию.

Ниже приведу пример как можно сделать directoryMustBeOpened, чтобы он проверял то что нужно с учётом вышеописанных проблем.

```java
    private static final String CLEAR_CONSOLE_COMMAND = "\033\\[H\033\\[J";

    @BeforeEach
    void prepareTmpDirectory() throws IOException {
        var tmpPath = Path.of("tmp");
        var fileSystemManagement = new FileSystemManagement();
        fileSystemManagement.delete(tmpPath.toString(), true);

        Files.createDirectory(tmpPath);
    }

    @Test
    public void directoryMustBeOpened() throws IOException {
        // папку tmp пришлось руками в коде установить как директорию по умолчанию
        Files.createDirectories(Path.of("tmp/dir")); 

        var commands = List.of("open dir", "finish");
        var commandBytes = String.join("\n", commands).getBytes();

        // в данном месте берём контроль над потоками ввода/выводы,
        // чтобы мы могли сымитировать ввод нужной команды и вычитать результат System.out.println
        System.setIn(new ByteArrayInputStream(commandBytes));
        var byteArrayOutputStream = new ByteArrayOutputStream();
        System.setOut(new PrintStream(byteArrayOutputStream));

        consoleTextFileEditor.start();
        
        // здесь дополнительно сносим из  вывода команду на очистку, чтобы не мешала нам
        var consoleOutput = byteArrayOutputStream.toString().replaceAll(CLEAR_CONSOLE_COMMAND, "");

        // если ожидаемый вывод получается очень большой на много строк, то можно 
        // сделать папку например testConsoleOutput и там положить текстовые файлы с большим выводом,
        // а в тесте уже сравнивать то что лежит в файле с реальным выводом
        Assertions.assertEquals("-d dir\r\nCurrent path: tmp\\dir\r\n", consoleOutput);
    }
```    
Данный принцип нужно использовать на все тесты по части тестирования ConsoleTextFileEditor.

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+    @Test
+    public void directoryMustNotBeDeletedBecauseDirectoryIsNotEmpty() {
+        manager.getDirectory(".\\src");
+        manager.createDirectory("p1");
```
В тестах необязательно использовать собственный код, в данном случае ты не проверяешь логику создания директории, поэтому вполне можно создать через File или Files, а для createDirectory будут уже свои тесты.

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+        Assertions.assertEquals("File newFile does not exist.", message);
+    }
+
+    private long filesCompareByByte(Path path1, Path path2) throws IOException {
```
Побайтовое сравнение не во всех случаях окей, бывают ситуации, когда текстовые файлы по содержимому идентичны, но в байтах есть незначительные различия. В случае текстового редактора я бы делал проверку текста по строкам.

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+        consoleTextFileEditor.delete(".\\src\\newFile");
+        consoleTextFileEditor.delete(".\\src\\newFile2");
+
+        Assertions.assertNotEquals(-1, res);
```
При такой проверке сложно будет понять, что будет означать отрицательный результат, тут либо использовать перегрузку и добавить сообщение, что это речь про разницу байт в сравнение или сделать assertTrue, тогда конструкция `test: fileMustBeEdited, result: false` уже понятней.

А так по идее нужно сравнивать строчки в файлах, тут всё очевидно будет.

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+
+        manager.delete(".\\src\\newFile.txt", false);
+
+        Assertions.assertTrue(absolutePath);
```
вместо assertTrue лучше использовать assertEquals, если тест упадёт будет очевидней какие данные неправильные.

```java
Assertions.assertEquals(Path.of("src/newFile.txt").toAbsolutePath(), map.get("Absolute path: "));
Assertions.assertEquals("8 bytes", map.get("Size: "));
```

Вообще у меня тест падает из-за
```
Expected :C:\Users\adgam\IdeaProjects\hw1_2\src\newFile.txt
Actual   :C:\Users\adgam\IdeaProjects\hw1_2\.\src\newFile.txt
```
у тебя точка залезла при формирование абсолютного пути, это баг по идеи

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditorTest.java:
```diff
@@ -0,0 +1,325 @@
+
+        manager.delete(".\\src\\newDirectory", false);
+
+        Assertions.assertTrue(absolutePath);
```
не хватает проверки на кол-во вложенных элементов

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+
+public class FileSystemManagement {
+
+    private File currentDirectory = new File(".");
```
Не хватает установки начальной директории из вне

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+    public FileSystemResponse<StringBuilder> getFile(String name) {
+        String path = currentDirectory.getPath() + "\\" + name;
+        File file = new File(path);
+        if (isDirectoryNotExists(path)) {
```
Можно обойтись без метода, используя `Files.notExists(Path.of(path))`

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+
+    private File currentDirectory = new File(".");
+
+    public FileSystemResponse<StringBuilder> getFile(String name) {
```
Я так понимаю StringBuilder был выбран из-за проблем с перегрузкой? Т.к. если использовать String, то используется конструктор ошибки? 

Я бы здесь всё таки строку отдавал, т.к. непонятно зачем во мне билдер, который к тому же ещё можно изменить.
Также ещё метод называется getFile, а получаем билдер, что вводит в заблуждение, метод я бы назвал что-то вроде `getTextFileContents`.

Над проблемой с перегрузкой предлагаю ещё подумать.

ml-rdv replied:
@Adgam Если возвращать Object? То есть return new FileSystemResponse<>((Object) sb.toString())

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+            return new FileSystemResponse<>("File " + file.getName() + " does not exist.");
+        }
+        StringBuilder sb = new StringBuilder();
+        if (extensionIsCorrect(file)) {
```
расширение файле можно проверить `file.getPath().endsWith(".txt")` без метода

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+                return new FileSystemResponse<>(e.toString());
+            }
+        }
+        return new FileSystemResponse<>("The file extension is incorrect.");
```
есть опасность , что выше будет добавлен ещё, что может привести неожиданному поведение. 
Здесь либо явно else прописать или вспоминай про принцип сначала валидация - потом основная логика

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+                }
+                return new FileSystemResponse<>(sb);
+            } catch (IOException e) {
+                e.printStackTrace();
```
В реальных приложениях вываливать StackTrace не очень хорошо клиенту, понятные ошибки всегда лучше + с точки зрения безопасности не показываем кишки приложения.

ml-rdv replied:
@Adgam fixed (remove e.printStackTrace())

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return new FileSystemResponse<>("The file extension is incorrect.");
+    }
+
+    private boolean extensionIsCorrect(File fileDirectory) {
```
название метода не отражает сути, он не просто проверяет корректность расширения, а проверяет, что файл текстовый
`boolean isTextFile(File file)` или 
`boolean isFileWithTextExtension(File file)`

ml-rdv replied:
@Adgam fixed (use file.getPath().endsWith(".txt") instead extensionIsCorrect)

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return extension.equals(".txt");
+    }
+
+    public FileSystemResponse<List<String>> getDirectory(String path) {
```
Метод делает 2 вещи:
1. устанавливает текущую директорию
2. отдаёт список содержимого текущей директории

Название метода getDirectory не отражает этих действий.

Предлагаю разделить на 2 метода.

ml-rdv replied:
@Adgam fixed

Adgam replied:
доработать

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return new FileSystemResponse<>(list);
+    }
+
+    public FileSystemResponse<Boolean> createFile(String name, String fileText) {
```
В методе большая вложенность и непонятно почему если метод fillOutFile выбрасывает исключение, то мы возвращаем валидный ответ. Также непонятно true будет только если пришёл текст, а  если без текста, то Unknown error.


```java
    public FileSystemResponse<Boolean> createFile(String name, String fileText) {
        File path = new File(currentDirectory.getPath() + "\\" + name);
        try {
           // избавляемся от вложенности за счёт инверсии + убрали блок try
            if (!path.createNewFile()) {
                return new FileSystemResponse<>("File with this name already exists.");
            }
            // доп проверка не пустую строку
            if (fileText != null && !fileText.isEmpty()) {
                fillOutFile(path.getPath(), fileText, false);
            }
            // вернём  true даже если текста нет
            return new FileSystemResponse<>(true);
        // здесь перехватываем все исключения и в том числе от fillOutFile
        } catch (Exception e) {
            return new FileSystemResponse<>(e.toString());
        }
```

P.S. увидел, что из ConsoleTextFileEditor если текста нет, то придёт пустая строка, поэтому такая логика. Проблема такого подхода в том, что мы завязываемся на внешнем классе, о котором по идеи ничего знать не должны, также это не очевидно, что всегда должна приходить пустая строка и поэтому будет возвращаться true.

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return new FileSystemResponse<>("Unknown error");
+    }
+
+    private void fillOutFile(String path, String fileText, boolean append) throws RuntimeException {
```
не уверен, что в этом методе много смысла

его можно свести к виду 
```java
  private void fillOutFile(String path, String fileText, boolean append) throws Exception {
        try (FileWriter output = new FileWriter(path, append)) {
            output.write(fileText);
        }
    }
```
но в таком случае такую конструкцию сразу можно использовать в createFile и в editFile

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    public FileSystemResponse<Boolean> back() {
+        if (currentDirectory.getPath().equals(".")) {
```
Жёсткая привязка к директории из которой запускаем приложение

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return new FileSystemResponse<>(directory.mkdir());
+    }
+
+    public FileSystemResponse<Boolean> back() {
```
название метода не раскрывает, что делает метод по факту

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        directory.delete();
+    }
+
+    public FileSystemResponse<Boolean> delete(String path, boolean deleteWithNestedDirectories) {
```
предлагаю метод разбить на 2

```java
   public FileSystemResponse<Boolean> delete(String path) {
        File directory = new File(path);
        if (isDirectoryNotExists(path)) {
            return new FileSystemResponse<>("Directory or file " + directory.getName() + " does not exist.");
        }
        File[] list = directory.listFiles();
        if (list == null || list.length == 0) {
            directory.delete();
            return new FileSystemResponse<>(true);
        }
        return new FileSystemResponse<>("Directory is not empty.");
    }

    public FileSystemResponse<Boolean> deleteWithNestedDirectories(String path) {
        File directory = new File(path);
        if (isDirectoryNotExists(path)) {
            return new FileSystemResponse<>("Directory or file " + directory.getName() + " does not exist.");
        }

        deleteDirectory(directory);
        return new FileSystemResponse<>(true);
    }
```

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        }
+        File newNameDirectory = new File(currentDirectory.getPath() + "\\" + newName);
+        FileSystemResponse<Boolean> result = new FileSystemResponse<>(oldNameDirectory.renameTo(newNameDirectory));
+        if (result.getBody()) {
```
так не проще будет?

```java
        if (oldNameDirectory.renameTo(newNameDirectory)) {
            return new FileSystemResponse<>(true);
        } else {
            return new FileSystemResponse<>("Directory or file with this name already exists.");
        }
````
а то не совсем понял, зачем нам создавать валидный ответ, если операция не выполнилась

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+            creationTimeFileTime = (FileTime) Files.getAttribute(file, "creationTime");
+        } catch (IOException e) {
+            e.printStackTrace();
+            return new FileSystemResponse<>(e.toString());
```
в данном случае, думаю отдать частичный ответ лучше, чем в ошибку упасть.
Если время создание получить не удалось, то можно проигнорировать или так и прописать, что не удалось определить.

ml-rdv replied:
@Adgam fixed

Adgam replied:
доработать

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+            e.printStackTrace();
+            return new FileSystemResponse<>(e.toString());
+        }
+        LocalDateTime localDateTime = creationTimeFileTime != null ? creationTimeFileTime
```
можно занести под if, чтобы убрать тернарки и проверки на null

```java
        if (creationTimeFileTime != null) {
            LocalDateTime localDateTime = creationTimeFileTime
                                .toInstant()
                                .atZone(ZoneId.systemDefault())
                                .toLocalDateTime();
            
            String creationTime = localDateTime.format(DateTimeFormatter.ofPattern("MM/dd/yyyy HH:mm"));
            
            info.put("Creation time: ", creationTime);
        }
```

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return new FileSystemResponse<>(info);
+    }
+
+    private int countNestedElements(File directory, int count) {
```
передача -1 как исходного значения путает, а ещё непонятно, что если передадим например 100, это что будет значить?

можно сделать проще 

```java
    int countNestedElements(File directory) {
        File[] files = directory.listFiles();
        var count = files != null ? files.length : 0;
        if (files != null) {
            for (var file: files) {
                count += count(file);
            }
        }
        return count;
    }
```

так или иначе, это не совсем то, что нужно. У папок кол-во вложенных элементов по умолчанию считается по вложенным элементам 1 уровня, проходится рекурсивно по всем-всем не надо было) Метод можешь оставить, просто в info название свойства изменить на более соответствующее.

ml-rdv replied:
@Adgam fixed
Вопрос. Что значит в инфо название свойства (Number of nested elements) изменить на более соответствующее? Изменить на "Number of elements"?

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -0,0 +1,230 @@
+        return count + 1;
+    }
+
+    private boolean isDirectoryNotExists(String path) {
```
метод проверяет и файлы, и деректории на сущестование и название немного сбивает с толку, лучше было бы назвать `isFileNotExists`, т.к. в java File это общая абстракция для файлов и директорий, ну и UNIX подобных системах действует правило "всё является файлом"

ml-rdv replied:
@Adgam fixed (use Files.notExists(Path.of(path)) instead isDirectoryNotExists)

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+public class ConsoleTextFileEditor {
+    private final FileSystemManagement manager = new FileSystemManagement();
+
+    public File getCurrentDirectory() {
```
здесь описал проблему https://github.com/ml-rdv/hw1_2/pull/9/files#r1486287720

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+        return manager.getCurrentDirectory();
+    }
+
+    public void start() {
```
кстати здорово было бы иметь метод что-то вроде `shutdown`,  который бы умел прерывать работу бесконечного цикла, в тестах бы очень пригодился бы, да и с точки зрения контракта не вижу проблем, раз `start` есть.

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+        String input;
+        while (true) {
+            input = in.nextLine();
+            System.out.print("\033[H\033[J");
```
вот этих ребят в метод бы вынести, что-то вроде 

```java
    private void clearConsole() {
        System.out.print("\033[H\033[J");
        System.out.flush();
    }
```
а то выглядят очень странно и непонятно, ещё не ясно что они делают

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+            input = in.nextLine();
+            System.out.print("\033[H\033[J");
+            System.out.flush();
+            if (!input.equals("finish")) {
```
инвертировать мб? или мб вообще finish вынести к обработке других команд и вызывать метод shutdown?

ml-rdv replied:
@Adgam fixed (invert finish)

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    public void continueWork(String input) {
```
название метода непонятно, что делает метод?

ml-rdv replied:
@Adgam fixed (rename method)

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+                String path = manager.getCurrentDirectory().getPath() + "\\" + splittedInput[1];
+                File directory = new File(path);
+                if (directory.isFile()) {
+                    openFile(splittedInput[1]);
```
splittedInput[1] хорошо вынести в переменную, т.к. что лежит по индексу 1 в splittedInput совсем непонятно

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+                if (splittedInput[1].equals("-f")) {
+                    String fileText = "";
+                    if (splittedInput.length > 3) {
+                        int dotIndex = input.lastIndexOf(splittedInput[3]);
```
что значит dotIndex?

ml-rdv replied:
@Adgam fixed (rename)

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+                        fileText = input.substring(dotIndex);
+                    }
+                    success = createFile(splittedInput[2], fileText);
+                    System.out.printf((success ? MESSAGE_FORMAT_SUCCESS : MESSAGE_FORMAT_NOT_SUCCESS),
```
унеси эти сообщение в FileSystemManagement и возвращай текст вместо boolean сразу, тогда уйдут тернарники + можешь избавиться от MESSAGE_FORMAT_SUCCESS и MESSAGE_FORMAT_NOT_SUCCESS, не вижу большого профита от того что они вынесены в константы. А вот понимание, что в итоге в строке отправляется затруднено.

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+                back();
+            }
+            case "edit" -> {
+                if (splittedInput.length > 2) {
```
сможешь свести к тому, что у всех команд будет просто вызов 1 метода как у back и info? метод continueWork сейчас делаем много всего, хорошо от него добиться выполнения одной задачи

ml-rdv replied:
@Adgam fixed

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+        System.out.println("Current path: " + manager.getCurrentDirectory().getPath());
+    }
+
+    private void back() {
```
много копипасты по части обработки FileSystemResponse, начиная с метода back() и до printInfo() почти всё тоже самое, подумай как уйти от копипасты

ml-rdv replied:
@Adgam fixed (add methods checkIsError, getMessageResult)

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+
+    public boolean delete(String path) {
+        FileSystemResponse<Boolean> fileSystemResponse = manager.delete(path, false);
+        if (fileSystemResponse.getMessageError() != null ||
```
тут как будто что-то с условием не то
разве не так должно быть `if (fileSystemResponse.getMessageError() != null && !fileSystemResponse.getMessageError().equals("Directory is not empty."))`?

тут ещё проблема, что завязываемся на сообщение "Directory is not empty.", как только оно измениться, то логика в консоли сломается

ml-rdv replied:
@Adgam Если в методе delete класса FileSystemManagement возвращать не Boolean, a Map<String, Boolean>, в которую помещать ("Directory is not empty", true), и в методе delete класса ConsoleTextFileEditor проверять, что если fileSystemResponse.getBoty содержит эту пару, то т.д.?

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/ConsoleTextFileEditor.java:
```diff
@@ -0,0 +1,233 @@
+        String input;
+        while (true) {
+            input = in.nextLine();
+            if (input.equals("Yes")) {
```
equalsIgnoreCase будет удобно с точки зрения UX

ml-rdv replied:
@Adgam fixed

ml-rdv committed:
Add .idea, MANIFEST.MF, SerializedObject to ignore

ml-rdv committed:
Add method clearConsole
Invert finish
Rename method continueWork to processCommand
Set name for splittedInput[1] in method processCommand
Rename dotIndex to indexFileText

ml-rdv committed:
Add method checkIsError, getMessageResult
Invert delete
Replace equals to equalsIgnoreCase
Replace tests
Add MESSAGE_FORMAT_SUCCESS/NOT to FileSystemManagement

ml-rdv committed:
Separation ConsoleTextFileEditorTest and FileSystemManagementTest
Moving initialization of consoleTextFileEditor and manager to tests

ml-rdv committed:
Comparing by lines instead by bytes
Add boolean filesAreSame

ml-rdv committed:
assertEquals instead assertTrue in test assertTrue
Add check for number of nested elements

ml-rdv committed:
Add opportunity to set start directory

ml-rdv committed:
Rename getFile to getTextFileContents

ml-rdv committed:
Use Files.notExists(Path.of(path))

ml-rdv committed:
Use file.getPath().endsWith(".txt") instead extensionIsCorrect method

ml-rdv committed:
Add else in getTextFileContents method for error because of file extension

ml-rdv committed:
Separation getDirectory to getDirectory and getListFiles

ml-rdv committed:
Change countNestedElements method

ml-rdv committed:
Add if for creationTimeFileTime in getInfo method

ml-rdv committed:
return new FileSystemResponse<>("Could not determine creation time");
instead
return new FileSystemResponse<>(e.toString());

move oldNameDirectory.renameTo(newNameDirectory) into if statement

ml-rdv committed:
return new FileSystemResponse<>("Could not determine creation time");
instead
return new FileSystemResponse<>(e.toString());

move oldNameDirectory.renameTo(newNameDirectory) into if statement

Separation delete method to delete and deleteWithNestedDirectories

ml-rdv committed:
Rename back to backToParentDirectory
Use currentDirectory.getParent() == null instead currentDirectory.getPath().equals(".") in if statement

ml-rdv committed:
Delete fillOutFile and move method body into createFile and editFile

ml-rdv committed:
Change in createFile method

ml-rdv committed:
Corrections

ml-rdv committed:
Use Files.create and Files.delete

ml-rdv committed:
Simplification processCommand method

ml-rdv committed:
Deleting setDirectory method
Add constructor for ConsoleTextFileEditor and FileSystemManagement
Add "Number of elements" (not nested) in getInfo method

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -12,12 +12,17 @@
 
-    public void setCurrentDirectory(File currentDirectory) {
-        this.currentDirectory = currentDirectory;
+    public FileSystemManagement(String directory) {
```
Можно ещё добавить конструктор 

```java
public FileSystemManagement() {
       this.currentDirectory = new File(DEFAULT_DIR);
}
```
чтобы не писать `new FileSystemManagement("")`, в данном случае непонятно что значит пустая строка.

ml-rdv replied:
@Adgam fixed ✅

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -12,12 +12,17 @@
+        if (Files.exists(Path.of(directory))) {
+            this.currentDirectory = new File(directory);
+        } else {
+            String DEFAULT_DIR = "";
```
CONSTANT_CASE в названиях переменных не помню чтобы где-то использовался, поэтому либо вынести как константу и оставить название или изменить на camelCase

ml-rdv replied:
@Adgam fixed ✅

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -181,22 +186,24 @@ public FileSystemResponse<Map<String, String>> getInfo(String name) {
-            info.put("Number of nested elements: ", String.valueOf(numberNestedElements));
+            info.put("Number of all nested elements: ", String.valueOf(numberNestedElements));
+
+            File dir = new File(directory.getPath());
```
у тебя уже есть переменная `directory`, зачем создавать новую?

ml-rdv replied:
@Adgam fixed ✅

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagement.java:
```diff
@@ -181,22 +186,24 @@ public FileSystemResponse<Map<String, String>> getInfo(String name) {
+        } catch (IOException e) {
+            info.put("Creation time: ", "Could not determine creation time");
         }
         if (directory.isDirectory()) {
```
проверка немного избыточная, listFiles() у файла вернёт null, а дальше у тебя уже стоят проверки, что если null, то 0 элементов

ml-rdv replied:
@Adgam idea сильно хотела, чтобы я поставила эту проверку и подчеркивала этот момент, поэтому добавила. Убрала ✅

Adgam replied:
я не учёл, что под условием ещё кладётся в hashMap кол-во элементов для директории и что по требованиям для файлов не должно быть такого свойства, поэтому тут скорей я не прав оказался)

ml-rdv committed:
Corrections: DEFAULT_DIR is constant
+ constructor without parameters for FileSystemManagement()

ml-rdv committed:
Добавила статический метод createErrorResponse() в класс FileSystemResponse, который вызывает приватный конструктор и инициализирует messageError
Это позволило избавиться от StringBuilder в методе getTextFileContents()

Переименовала getDirectory() в setDirectory()

Разбила методы setDirectory() и getListFiles()

Исправила if (!Files.notExists(Path.of(path))) на if (Files.exists(Path.of(path))) в методе createDirectory

ml-rdv committed:
Correct tests, those check interaction between console and system in / out.

ml-rdv committed:
Add test shouldPrintInfoAboutDirectory() and shouldPrintCommandIsNotCorrect() to ConsoleTextFileEditorTest

Delete static fields MESSAGE_FORMAT_SUCCESS and MESSAGE_FORMAT_NOT_SUCCESS

Return FileSystemResponse<String> with text of success result / not success instead returning FileSystemResponse<Boolean>

ml-rdv committed:
Check logics in FileSystemManagementTest

Adgam commented on src/InputOutputStream/ConsoleTextFileEditor/FileSystemManagementTest.java:
```diff
@@ -85,9 +87,12 @@ public void fileMustBeCreated() throws IOException {
         Assertions.assertEquals("File newFile has been successfully created.\n", messageBody);
+        Assertions.assertTrue(Files.exists(path));
+
+        Files.delete(path);
```
а зачем вот вот эти удаления? у тебя же папка "work" пересоздаётся перед каждым тестом
